Tools/cases_generator/generate_cases.py

```diff
@@ -165,122 +215,101 @@ def write_body(self, f: typing.TextIO, ndent: str, dedent: int) -> None:
                     else:
                         break
                 if ninputs:
-                    f.write(f"{space}if ({cond}) goto pop_{ninputs}_{label};\n")
+                    out.write_raw(f"{space}if ({cond}) goto pop_{ninputs}_{label};\n")
                 else:
-                    f.write(f"{space}if ({cond}) goto {label};\n")
+                    out.write_raw(f"{space}if ({cond}) goto {label};\n")
             else:
-                f.write(line)
+                out.write_raw(line)
+
+
+InstructionOrCacheEffect = Instruction | parser.CacheEffect
 
 
 @dataclasses.dataclass
-class SuperComponent:
+class Component:
     instr: Instruction
     input_mapping: dict[str, parser.StackEffect]
     output_mapping: dict[str, parser.StackEffect]
 
+    def write_body(self, out: Formatter, cache_adjust: int) -> None:
+        with out.block(""):
+            for var, ieffect in self.input_mapping.items():
+                out.emit(f"PyObject *{ieffect.name} = {var};")
+            for oeffect in self.output_mapping.values():
+                out.emit(f"PyObject *{oeffect.name};")
+            self.instr.write_body(out, dedent=-4, cache_adjust=cache_adjust)
+            for var, oeffect in self.output_mapping.items():
+                out.emit(f"{var} = {oeffect.name};")
+
 
-class SuperInstruction(parser.Super):
+# TODO: Use a common base class for {Super,Macro}Instruction
 
+
+@dataclasses.dataclass
+class SuperOrMacroInstruction:
+    """Common fields for super- and macro instructions."""
+
+    name: str
     stack: list[str]
     initial_sp: int
     final_sp: int
-    parts: list[SuperComponent]
-
-    def __init__(self, sup: parser.Super):
-        super().__init__(sup.kind, sup.name, sup.ops)
-        self.context = sup.context
-
-    def analyze(self, a: "Analyzer") -> None:
```

Comment:
All this analysis code has been moved into class `Analyzer`, e.g. `analyze_supers_and_macros`.